nixos/modules/services/torrent/deluge.nix

```diff
@@ -20,45 +44,143 @@ in {
+          default = "/var/lib/deluge";
+          description = ''
+            The directory where deluge will create files.
+          '';
+        };
+
+        authFile = mkOption {
+          type = types.path;
+          example = "/run/keys/deluge-auth";
+          description = ''
+            The file managing the authentication for deluge, the format of
+            this file is straightforward, each line contains a
+            username:password:level tuple in plaintext.
+            Only has an effect when <option>services.deluge.declarative</option>
+            is set to <literal>true</literal>.
+          '';
+        };
       };
 
-      deluge.web.enable = mkEnableOption "Deluge Web daemon";
+      deluge.web = {
+        enable = mkEnableOption "Deluge Web daemon";
+
+        port = mkOption {
+        type = types.ints.u16;
+          example = 8112;
+          description = ''
+            Deluge web UI port.
+          '';
+        };
+      };
     };
   };
 
   config = mkIf cfg.enable {
 
+    systemd.tmpfiles.rules = [ "d '${configDir}' 0700 deluge deluge" ]
+    ++ optional (hasAttr "download_location" cfg.config) "d '${cfg.config.download_location}' 0770 deluge deluge"
+    ++ optional (hasAttr "torrentfiles_location" cfg.config) "d '${cfg.config.torrentfiles_location}' 0770 deluge deluge"
+    ++ optional (hasAttr "move_completed_path" cfg.config) "d '${cfg.config.move_completed_path}' 0770 deluge deluge";
+
     systemd.services.deluged = {
       after = [ "network.target" ];
       description = "Deluge BitTorrent Daemon";
       wantedBy = [ "multi-user.target" ];
       path = [ pkgs.deluge ];
       serviceConfig = {
-        ExecStart = "${pkgs.deluge}/bin/deluged -d";
+        ExecStart = ''
+          ${pkgs.deluge}/bin/deluged \
+            --do-not-daemonize \
+            --config ${configDir}
+        '';
         # To prevent "Quit & shutdown daemon" from working; we want systemd to manage it!
         Restart = "on-success";
         User = "deluge";
         Group = "deluge";
+        UMask = "0002";
         LimitNOFILE = cfg.openFilesLimit;
       };
+      preStart = preStart;
     };
 
     systemd.services.delugeweb = mkIf cfg_web.enable {
-      after = [ "network.target" ];
+      after = [ "network.target" "deluged.service"];
+      requires = [ "deluged.service" ];
       description = "Deluge BitTorrent WebUI";
       wantedBy = [ "multi-user.target" ];
       path = [ pkgs.deluge ];
-      serviceConfig.ExecStart = "${pkgs.deluge}/bin/deluge --ui web";
-      serviceConfig.User = "deluge";
-      serviceConfig.Group = "deluge";
+      serviceConfig = {
+        ExecStart = ''
+          ${pkgs.deluge}/bin/deluge-web \
+            --config ${configDir} \
+            ${optionalString (cfg.web.port != null) "--port ${toString cfg.web.port}"}
+        '';
+        User = "deluge";
+        Group = "deluge";
+      };
+    };
+
+    networking.firewall = mkIf cfg.openFirewall {
```

Comment:
Since this can only work when `declarative` is enabled, this should be `mkIf (cfg.declarative && cfg.openFirewall`)`. And in the description of `openFirewall` you should also mention this.